### DISCUSSION TITLE
Can you explain why headless requires a sleep for the same code that does not from the UI?

### DISCUSSION THREAD

**Discussion by Vasco-jofra** [Vector35/binaryninja-api#2296] [Q&A] [answered]
As seen in the image below, iterating and printing the instructions of a newly created function does not work, unless you sleep between creating the function and printing its instructions.

<img width="517" alt="Screenshot 2021-03-04 at 12 43 59" src="https://user-images.githubusercontent.com/11303847/109965803-5fb63b80-7ce7-11eb-99aa-2790a24c3e8f.png">

The code:
```python
from binaryninja import BinaryView, Function, Platform
import time

# Create a binary view from a data buffer
new_bv: BinaryView = BinaryView.new("\x55\x55\x90\x90\xc3")
new_bv.platform = Platform["linux-x86_64"]

# Create a function at the start address
new_bv.create_user_function(0)
func: Function = new_bv.get_function_at(0)

time.sleep(0.1) # <--- remove me and the print below does not work

for instr in func.instructions:
    print(instr)
```

**Comment by bpotchik** [comment, marked as answer]
In headless you should use  [`new_bv.update_analysis_and_wait()`](https://api.binary.ninja/binaryninja.binaryview-module.html#binaryninja.binaryview.BinaryView.update_analysis_and_wait) instead of `sleep`.

This is different than the scripting console in the full UI because in that case, we assume you want to see updated results immediately for any command you run as it violates the principal of least surprise. You can see that here:

https://github.com/Vector35/binaryninja-api/blob/dev/python/scriptingprovider.py#L676-L677

Headless however is much more likely to be building a set of actions and you may want more control over when analysis updates since you can queue up multiple actions without being able to see the results.

**Comment by Vasco-jofra** [comment]
Thanks, `new_bv.update_analysis_and_wait()` works. I though that after calling `create_user_function`, the fields of that function (like `func.instructions`) would be ready to use. 

If this is intended we can just close it

**Comment by bpotchik** [comment]
Ah yes, `update_analysis_and_wait()`, corrected above.

So for headless processing it make sense to batch multiple analysis actions before requesting the core to perform the analysis. This can yield better performance, especially on multi-core machines. Thanks!
